README.md
# img
## [pub.dev Listing](https://pub.dev/packages/img) | [API Doc](https://pub.dev/documentation/img/latest) | [GitHub](https://github.com/Zabadam/img)
#### API References: [`Repeat`](https://pub.dev/documentation/img/latest/img/Repeat-class.html) | [`ImageToo`](https://pub.dev/documentation/img/latest/img/ImageToo-class.html) | [`ImageDecorationToo`](https://pub.dev/documentation/img/latest/img/ImageDecorationToo-class.html) | [`paintImageToo()`](https://pub.dev/documentation/img/latest/img/paintImageToo.html) | [`InkImg`](https://pub.dev/documentation/img/latest/img/InkImg-class.html)

## 🙋‍♂️ I'm an `Image` Too!
Easily paint mirror-tiling images for seamless edge-to-edge textures from
any source. Defines Repeat, an expansion on ImageRepeat that enables
mirroring, as well as the Widgets, extensions, and methods to support it.

An `ImageToo` or `DecorationImageToo` is used to render images from the
expected array of sources, but their `repeat` property is expanded.

The secret sauce is the [`Repeat`](https://pub.dev/documentation/img/latest/img/Repeat-class.html)
enum with the standard options from `ImageRepeat`, such as `noRepeat` and
`repeatX`, as well as bespoke mirror values, spanning `mirrorX`, `mirrorY`,
and global `mirror`.

[![animation demonstrating the example app](https://raw.githubusercontent.com/Zabadam/img/main/doc/example.gif)](https://pub.dev/packages/img/example 'animation demonstrating the example app')

This package forks several of Flutter's vanilla classes and `Widget`s,
rigging them to drive an alternate paint method that is the real meat and
potatoes of 🙋‍♂️ [`img`](https://pub.dev/packages/img).

Predictably named [`paintImageToo()`](https://pub.dev/documentation/img/latest/img/paintImageToo.html),
this method performs mostly the same as the built-in `paintImage()` method,
but it does make a few considerations for the repeat styles.
- Namely, if you are interested, within the standard tile generation method
  employed for `ImageRepeat.repeat`, an extra pair of values is provided back
  to the main painting body.
  - These "proximities" are values whose parity (even/oddness) dictate whether
  an image needs to be mirrored before painting. More specifically, the canvas
  itself is mirrored, the image is painted, then the canvas is restored.
  - The original source tile image would be `Proximity(0,0)`, and the `Rect`
  directly to its right would have `Proximity(1,0)`.
 
Some images fare better than others as far as the quality of the output.
Most will appear kaleidoscopic at worst and magical at best.
 
Still, condiering how *few* images out there are designed to support
edge-to-edge tiling with a simple `ImageRepeat.repeat` mode, this functionality
broadly expands the versatility of using any image as a seamless texture.

## Getting Started
To place an image directly into an application *as a `Widget`*, employ a 
`new ImageToo(. . .)`.

```dart
void main() => runApp(const Example());

class Example extends StatelessWidget {
  const Example({Key? key}): super(key: key);
  
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        backgroundColor: Colors.black,
        body: Center(
          child: ImageToo(
            image: NetworkImage(
              'https://gifimage.net/wp-content/uploads/2017/08/transparent-fire-gif-22.gif',
              scale: 1,
            ),
            repeat: Repeat.mirror,
            width: 600,
            height: 600,
          ),
        ),
      ),
    );
  }
}
```

|                                                                                                     Sample code output                                                                                                      |                                                                                              Original image (1x tile)                                                                                              |
| :-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: | :----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
| [![Sample code output](https://raw.githubusercontent.com/Zabadam/img/main/doc/readme_sample0.gif 'Sample code output, `repeat: Repeat.mirror`')](https://raw.githubusercontent.com/Zabadam/img/main/doc/readme_sample0.gif) | [![Original image (1x tile)](https://gifimage.net/wp-content/uploads/2017/08/transparent-fire-gif-22.gif 'Original image (1x tile)')](https://gifimage.net/wp-content/uploads/2017/08/transparent-fire-gif-22.gif) |

This shiny new `Repeat` with its mirror values can also be used to decorate a
`Container` or anywhere else `DecorationImage` might typically be placed.

Just swap out your vanilla `DecorationImage` object with a `DecorationImageToo`
object, replacing the `ImageRepeat repeat` property with a `Repeat repeatMode`.


```dart
void main() => runApp(const FloorIsLava());

class FloorIsLava extends StatelessWidget {
  const FloorIsLava({Key? key}) : super(key: key);

  Widget build(BuildContext context) => const MaterialApp(
        debugShowCheckedModeBanner: false,
        home: Scaffold(
          backgroundColor: Colors.black,
          body: ExampleBody(),
        ),
      );
}

class ExampleBody extends StatelessWidget {
  const ExampleBody({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Container(
        width: MediaQuery.of(context).size.width,
        height: MediaQuery.of(context).size.height,
        decoration: const BoxDecoration(
          color: Colors.amber,
          image: DecorationImageToo(
            image: NetworkImage(
              'https://gifimage.net/wp-content/uploads/2017/08/transparent-fire-gif-22.gif',
              scale: 14,
            ),
            repeatMode: Repeat.mirror,
          ),
        ),
        child: const Center(
          child: Text(
            'THE\nFLOOR\nIS\nLAVA',
            textAlign: TextAlign.center,
            style: TextStyle(
              fontSize: 75,
              color: Colors.black,
              fontWeight: FontWeight.bold,
            ),
          ),
        ),
      ),
    );
  }
}
```

| Sample code output                                                                                                                                                                                                          | Original image (1x tile)                                                                                                                                                                                           |
| :-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | :----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| [![Sample code output](https://raw.githubusercontent.com/Zabadam/img/main/doc/readme_sample1.gif 'Sample code output, `repeat: Repeat.mirror`')](https://raw.githubusercontent.com/Zabadam/img/main/doc/readme_sample1.gif) | [![Original image (1x tile)](https://gifimage.net/wp-content/uploads/2017/08/transparent-fire-gif-22.gif 'Original image (1x tile)')](https://gifimage.net/wp-content/uploads/2017/08/transparent-fire-gif-22.gif) |

## Advanced Usage

### Extension Methods
For your consideration is an extension for making a quick textured surface from
a single url as `String.toSeamlessTexture()`.

```dart
/// URL leading to a Tatsuro Yamashita album cover.
const tatsu = 'https://spice.eplus.jp/images/KefMrp9J1bM7NGRvFqK64ZNOfbTGUDKVCC8ePaiKB1cOcOJz1rEN3DQUJMBZhQJ2.jpg';

/// Extension method
final extensionExamples = <Widget>[
  tatsu.toSeamlessTexture(),
  tatsu.toSeamlessTexture(scale: 10),
  InteractiveViewer(
    maxScale: 150,
    child: tatsu.toSeamlessTexture(scale: 75),
  );
];
```

Expect output from these widgets to resemble something like this:

[![This code sample comes from the package example](https://raw.githubusercontent.com/Zabadam/img/main/doc/extensions.gif)](https://pub.dev/packages/img/example 'This code sample comes from the package example')

Beyond `'https://url.to/image.png'.toSeamlessTexture()`, consider:

```dart
'https://url.to/image.png'.toSeamlessRow();
'https://url.to/image.png'.toSeamlessColumn();

'res/image.gif'.toSeamlessTexture(isAsset: true, package: 'package_name');

File.toSeamlessColumn();
Uint8List.toSeamlessRow();
```

### 🙋‍♂️ `InkImg`
Extends `Ink` and has the same paramters as `Ink.image`, but creates an ink
decoration with a `DecorationImageToo` that supports `Repeat`ing (mirroring 😉).

Additionally, the original `Ink.image` does not pass a color to its decoration,
but `InkImg` will. This optional color is painted behind the image.

### ❗ Note

When using any of the new mirror modes as the `repeat` property, the
[`paintImageToo()`](https://pub.dev/documentation/img/latest/img/paintImageToo.html)
method will override the image's alignment to `Alignment.center`.

This is a workaround at the moment for invalid rendering when alignment is not
centered.

To compensate for now, an `Offset` property is available with values that range
`0..maxResolution` where `maxResolution` is the axis length in pixels of the
image for the corresponding offset component, `dx` or `dy`; but this only
*mimics* "alignment" with a `Repeat.mirror` tiled image that is meant to fill
an entire space.

That is, for an image with the resolution 600✖400, an `ImageToo` or
`DecorationImageToo` may have a tile shift maxing out at `Offset(600,400)`
(and defaulting to `Offset.none`) applied as its `mirrorOffset` property
when its repeat is set to one of the mirror modes.

Stay tuned if you would like to, say, align the image left and also
`Repeat.mirrorY`.

### Deep Dive

In order to dig deep enough to the painting method, several intermediate 
image "too" classes had to be forked. All told, from order of direct
developer relevance:

#### 🙋‍♂️ `ImageToo`
\- Shorthand: `Img`

A stateful widget, either `const` via `ImageProvider` or through a variety of named convenience constructors, that passes along the new `Repeat` value. Consider `ImageToo.asset` or `ImageToo.network`, etc.

#### 🙋‍♂️ `DecorationImageToo`
\- Shorthand: `DecorationImg`

A `Repeat`-conscious variant of a `Decoration.image`-fulfilling `DecorationImage`.

| <h4>🙋‍♂️ `RawImageToo`</h4> | <h4>🙋‍♂️ `RenderImageToo`</h4> |
| :----------------------: | :-------------------------: |

Not currently exported with the package. Could be. Should they be?

Dart abstraction and `LeafRenderObjectWidget` created by an `ImageToo` that
then creates the `RenderBox`.

## 🛣️ Roadmap
1. Fix `alignment` and likely remove `mirrorOffset`.

<br />

---

## 🐸 [Zaba.app ― simple packages, simple names.](https://pub.dev/publishers/zaba.app/packages 'Other Flutter packages published by Zaba.app')

<details>
<summary>More by Zaba</summary>

### Widgets to wrap other widgets
- ## 🕹️ [xl](https://pub.dev/packages/xl 'implement accelerometer-fueled interactions with a layering paradigm')
- ## 🌈 [foil](https://pub.dev/packages/foil 'implement accelerometer-reactive gradients in a cinch')
- ## 📜 [curtains](https://pub.dev/packages/curtains 'provide animated shadow decorations for a scrollable to allude to unrevealed content')
---
### Container widget that wraps many functionalities
- ## 🌟 [surface](https://pub.dev/packages/surface 'animated, morphing container with specs for Shape, Appearance, Filter, Tactility')
---
### Side-kick companions, work great alone or employed above
- ## 🙋‍♂️ [icon](https://pub.dev/packages/icon 'An extended Icon \"Too\" for those that are not actually square, plus shadows support + IconUtils')
- ## 🙋‍♂️ [img](https://pub.dev/packages/img 'An extended Image \"Too\" and DecorationImageToo that support an expanded Repeat.mirror painting mode')
- ## 🏓 [ball](https://pub.dev/packages/ball 'a bouncy, position-mirroring splash factory that\'s totally customizable')
- ## 👥 [shadows](https://pub.dev/packages/shadows 'convert a double-based \`elevation\` + BoxShadow and List\<BoxShadow\> extensions')
</details>


CHANGELOG.md
# img
## **[0.1.0]** - 07 JUN 21
- Initial release.


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:img/img.dart';

void main() => runApp(const ExampleFrame());

/// [MaterialApp] frame.
class ExampleFrame extends StatelessWidget {
  /// [MaterialApp] frame.
  const ExampleFrame({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) => const MaterialApp(
        debugShowCheckedModeBanner: false,
        home: Example(),
      );
}

/// Construct a [new Example] `Widget` to fill an [ExampleFrame].
class Example extends StatelessWidget {
  /// Fill an [ExampleFrame] with a [Scaffold] and [AppBar] whose body is a
  /// [PageView]. The `children` of this swiping page view are each built by
  /// [buildView], save for an example that is recreated from a README sample
  /// and another example demonstrating [StringToTexture].
  const Example({Key? key}) : super(key: key);

  /// One entire page for a [PageView]. Comprised of a [SingleChildScrollView]
  /// and a [Column], it presents a series of [Text] and [Img].
  Widget buildView({
    required Repeat repeat,
    required double w,
    required double h,
    required String url,
    required double scale1,
    required double scale2,
    required String mode,
    required String subtitle,
  }) =>
      SingleChildScrollView(
        child: Column(
          children: [
            const SizedBox(height: 25),
            ImageToo(image: NetworkImage(url, scale: scale1)),
            const SizedBox(height: 25),
            Text(
              'Repeat.$mode',
              style: const TextStyle(fontSize: 30, color: Colors.white),
            ),
            Text(
              subtitle,
              style: const TextStyle(fontSize: 20, color: Colors.white),
            ),
            Center(
              /// Equivalent to [ImageToo]
              child: Img(
                image: NetworkImage(url, scale: scale2),
                repeat: repeat,
                width: w,
                height: h,
              ),
            ),
          ],
        ),
      );

  @override
  Widget build(BuildContext context) {
    final s = MediaQuery.of(context).size;
    final w = s.width;
    final h = s.height;

    return Scaffold(
      backgroundColor: Colors.black,
      appBar: AppBar(title: const Text('img')),
      body: PageView(
        physics: const BouncingScrollPhysics(),
        children: [
          buildView(
            repeat: Repeat.noRepeat,
            w: w,
            h: h,
            url: window,
            scale1: 3,
            scale2: 1,
            mode: 'noRepeat',
            subtitle: '',
          ),
          buildView(
            repeat: Repeat.repeatX,
            w: w,
            h: h,
            url: fire,
            scale1: 3,
            scale2: 1,
            mode: 'repeatX',
            subtitle: 'Works great because this image naturally tiles',
          ),
          buildView(
            repeat: Repeat.repeat,
            w: w,
            h: h,
            url: fire,
            scale1: 3,
            scale2: 2,
            mode: 'repeat',
            subtitle: 'but uh oh... what if we need fire everywhere?',
          ),
          buildView(
            repeat: Repeat.mirror,
            w: w,
            h: h,
            url: fire,
            scale1: 3,
            scale2: 4,
            mode: 'mirror',
            subtitle: 'Now that\'s strange, but a little better!',
          ),
          buildView(
            repeat: Repeat.mirror,
            w: w,
            h: h,
            url: glitter,
            scale1: 3,
            scale2: 6,
            mode: 'mirror',
            subtitle: 'Turn any image into a seamless texture.',
          ),
          buildView(
            repeat: Repeat.mirror,
            w: w,
            h: h,
            url: frog,
            scale1: 3,
            scale2: 2,
            mode: 'mirror',
            subtitle: 'Or create a kalediscopic dreamworld!',
          ),

          /// Sample from README, adjusted
          Center(
            child: Container(
              width: w * 2 / 3,
              height: h * 2 / 3,
              decoration: const BoxDecoration(
                color: Colors.amber,
                image: DecorationImageToo(
                  image: NetworkImage(fire, scale: 14),
                  repeatMode: Repeat.mirror,
                ),
              ),
              child: const Center(
                child: Text(
                  'THE\nFLOOR\nIS\nLAVA',
                  textAlign: TextAlign.center,
                  style: TextStyle(
                    fontSize: 75,
                    color: Colors.black,
                    fontWeight: FontWeight.bold,
                  ),
                ),
              ),
            ),
          ),

          /// Extension methods
          tatsu.toSeamlessTexture(),
          tatsu.toSeamlessTexture(scale: 10),
          InteractiveViewer(
            maxScale: 150,
            child: tatsu.toSeamlessTexture(scale: 75),
          ),

          const LovelyRainyDay(),
        ],
      ),
    );
  }
}

/// Construct a [Stack] of mirror-tiling images formed by
/// `String.toSeamlessTexture` method with a [new LovelyRainyDay].
class LovelyRainyDay extends StatelessWidget {
  /// A [Stack] of mirror-tiling images formed by
  /// `String.toSeamlessTexture` method.
  ///
  /// Draws a scene of a lovely spring day with clouds, hills, a pond,
  /// and some rain.
  const LovelyRainyDay({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: [
        Positioned.fill(child: clouds.toSeamlessTexture(scale: 2, height: 300)),
        Positioned(
          bottom: 0,
          left: 0,
          right: 0,
          child: Container(
            child: field.toSeamlessRow(height: 300),
          ),
        ),
        for (var hill in hillLayers) hill,
        for (var shower in rainLayers) shower,
      ],
    );
  }
}

/// A group of [Positioned] images that are formed into mirroring textures
/// from simple `String`s.
final hillLayers = [
  Positioned(
    bottom: 300,
    left: 0,
    right: 0,
    child: hills.toSeamlessRow(
      height: 190,
      color: Colors.blueGrey[900]!.withOpacity(0.75),
      blendMode: BlendMode.srcATop,
    ),
  ),
  Positioned(
    bottom: 235,
    left: 0,
    right: 0,
    child: hills.toSeamlessRow(
      height: 190,
      color: Colors.black.withOpacity(0.35),
      blendMode: BlendMode.srcATop,
    ),
  ),
  Positioned(
    bottom: 180,
    left: 0,
    right: 0,
    child: hills.toSeamlessRow(
      height: 100,
      color: Colors.blueGrey[800]!.withOpacity(0.4),
      blendMode: BlendMode.srcATop,
    ),
  ),
  Positioned(
    bottom: 145,
    left: 0,
    right: 0,
    child: hills.toSeamlessRow(height: 50, offset: const Offset(20, 0)),
  ),
];

/// A group of [Positioned] images that are formed into mirroring textures
/// from simple `String`s.
final rainLayers = [
  Positioned(
    left: -450,
    top: 0,
    right: 0,
    bottom: 0,
    child: rain.toSeamlessColumn(scale: 2),
  ),
  Positioned(
    left: -400,
    top: 0,
    right: 0,
    bottom: 0,
    child: rain.toSeamlessColumn(scale: 3),
  ),
  Positioned(
    left: -225,
    top: 0,
    right: 0,
    bottom: 0,
    child: rain.toSeamlessColumn(scale: 5),
  ),
];

/// URL leading to an animated image of a window with a moon and tain.
const window = 'https://www.gifstop.com/images/weathergifs/raingif.gif';

/// URL leading to an animated image of fire.
const fire = 'https://gifimage.net/wp-content/uploads/2017/08/'
    'transparent-fire-gif-22.gif';

/// URL leading to an animated image colored glitter.
const glitter = 'https://i.giphy.com/media/R2B4OcBpN7YS4/giphy.gif';

/// URL leading to an animated image of a frog floating in water.
const frog = 'https://www.gifmania.it/Gif-Animate-Animali/'
    'Immagini-Animate-Rane/Gif-Animati-Immagini-Di-Rane/'
    'Immagini-Di-Rane-66777.gif';

/// URL leading to a Tatsuro Yamashita album cover.
const tatsu = 'https://spice.eplus.jp/images/'
    'KefMrp9J1bM7NGRvFqK64ZNOfbTGUDKVCC8ePaiKB1cOcOJz1rEN3DQUJMBZhQJ2.jpg';

/// URL leading to an image of clouds.
const clouds = 'https://i.pinimg.com/736x/9e/03/4d/'
    '9e034dec913971c7750e4836dc1f479b.jpg';

/// URL leading to an image of a field.
const field =
    'https://static.vecteezy.com/system/resources/previews/000/519/977/original/open-field-vector.jpg';

/// URL leading to an image of hills.
const hills = 'http://clipart-library.com/img1/1255972.png';

/// URL leading to an animated image of rain.
const rain = 'https://i.giphy.com/media/495IrnWutQBc4hhhoL/giphy.gif';
